main/src/ca/uwaterloo/flix/language/phase/njvm/GenRefClasses.scala

```diff
@@ -0,0 +1,114 @@
+/*
+ * Copyright 2019 Miguel Fialho
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package ca.uwaterloo.flix.language.phase.njvm
+
+import ca.uwaterloo.flix.api.Flix
+import ca.uwaterloo.flix.language.ast.FinalAst.Root
+import ca.uwaterloo.flix.language.phase.jvm.{JvmClass, JvmName, JvmType}
+import ca.uwaterloo.flix.language.phase.njvm.Mnemonics._
+import ca.uwaterloo.flix.language.phase.njvm.Mnemonics.Instructions._
+import ca.uwaterloo.flix.language.phase.njvm.Mnemonics.JvmModifier._
+
+
+/**
+  * Generates bytecode for the cell classes.
+  */
+object GenRefClasses1 {
+  // TODO: incrementally copy over.
+
+
+  def gen()(implicit root: Root, flix: Flix): Map[JvmName, JvmClass]  = {
+    // Type that we need a cell class for
+    val types = List(JvmType.PrimBool, JvmType.PrimChar, JvmType.PrimFloat, JvmType.PrimDouble,
+      JvmType.PrimByte, JvmType.PrimShort, JvmType.PrimInt, JvmType.PrimLong, JvmType.Object)
+
+    // Generating each cell class
+    types.map { tpe =>
+      val classType = JvmName.getCellClassType(tpe)
+      classType.name -> JvmClass(classType.name, genRefClass(classType, tpe))
+    }.toMap
+  }
+
+  /**
+    * Generating class `classType` with value of type `tpe`
+    */
+  def genRefClass(classType : JvmType.Reference, cellType : JvmType)(implicit root: Root, flix: Flix): Array[Byte] = {
+
+    val cg = new ClassGenerator(classType,List(Public,Final), JvmName.Object.toInternalName, null)
+
+    val field0 = cg.compileField(List(Private),"field0", cellType)
+    val constructorLocals = new FunSig1(List(Public), "<init>", JvmType.Void, classType, cellType)
```

Comment:
Presumably these could be methods on CG?

Reply:
We should probably have a special FunSig1 just for constructors (methods name <init>)?